src/gba/hardware-extensions.c

```diff
@@ -0,0 +1,339 @@
+#include <mgba/internal/gba/gba.h>
+
+
+enum HWEX_RETURN_CODES {
+    HWEX_RET_OK = 0,
+    HWEX_RET_WAIT = 1,
+
+    // Errors
+    HWEX_RET_ERR_UNKNOWN = 0x100,
+    HWEX_RET_ERR_BAD_ADDRESS = 0x101,
+    HWEX_RET_ERR_INVALID_PARAMETERS = 0x102,
+    HWEX_RET_ERR_WRITE_TO_ROM = 0x103,
+    HWEX_RET_ERR_ABORTED = 0x104,
+    HWEX_RET_ERR_DISABLED = 0x105,
+    HWEX_RET_ERR_DISABLED_BY_USER = 0x106
+};
+
+#define SIMPLIFY_HWEX_REG_ADDRESS(address) ((address - REG_HWEX0_CNT) >> 1)
+
+const uint16_t extensionIdByRegister[] = {
+    // More RAM
+    [SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_CNT)] = HWEX_ID_MORE_RAM,
+    [SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_RET_CODE)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P0_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P0_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P1_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P1_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P2_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P2_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P3_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P3_HI)] = HWEX_ID_MORE_RAM,
+};
+
+static uint16_t GetExtensionIdFromAddress(uint32_t address) {
+    return extensionIdByRegister[SIMPLIFY_HWEX_REG_ADDRESS(address)];
+}
+
+#undef SIMPLIFY_HWEX_REG_ADDRESS
+
+// CNT flags
+// Writable
+#define HWEX_CNT_FLAG_CALL 1
+#define HWEX_CNT_ALL_WRITABLE (HWEX_CNT_FLAG_CALL)
+// Read only
+#define HWEX_CNT_FLAG_PROCESSING (1 << 15)
+
+struct HardwareExtensionHandlers {
+    uint16_t (*onCall)(struct GBA* gba);
+    bool (*onAbort)(void);
+};
+
+static const struct HardwareExtensionHandlers extensionHandlers[];
+
+void GBAHardwareExtensionsInit(struct GBAHardwareExtensions* hwExtensions) {
+	hwExtensions->enabled = false;
+
+    hwExtensions->userEnabled = false;
+    memset(hwExtensions->userEnabledFlags, 0, sizeof(hwExtensions->userEnabledFlags));
+
+    memset(hwExtensions->memory, 0, sizeof(hwExtensions->memory));
+}
+
+static uint32_t GetHwExMemoryIndex16FromAddress(uint32_t address) {
+    return (address - REG_HWEX_ENABLE_FLAGS_0) >> 1;
+}
+
+static uint32_t GetHwExMemoryIndex32FromAddress(uint32_t address) {
+    return (address - REG_HWEX_ENABLE_FLAGS_0) >> 2;
+}
+
+static uint16_t* GetHwExIOPointer(struct GBA* gba, uint32_t address) {
+    return ((uint16_t*) &gba->hwExtensions.memory) + GetHwExMemoryIndex16FromAddress(address);
+}
+
+static uint16_t _GBAHardwareExtensionsIORead(struct GBA* gba, uint32_t address) {
+    switch (address) {
+        case REG_HWEX_ENABLE:
+            return 0x1DEA;
+        case REG_HWEX_VERSION:
+            return REG_HWEX_VERSION_VALUE;
+        case REG_HWEX_ENABLE_FLAGS_0:
+        case REG_HWEX_ENABLE_FLAGS_1:
+        case REG_HWEX_ENABLE_FLAGS_2:
+        case REG_HWEX_ENABLE_FLAGS_3:
+        case REG_HWEX_ENABLE_FLAGS_4:
+        case REG_HWEX_ENABLE_FLAGS_5: {
```

Comment:
What is the purpose of making these registers generic? Why not just have a fixed address per extension?

Reply:
No particular reason. This way they occupy 1 bit instead of 16. But yeah, I could change it so it works the same as REG_HWEX_ENABLE.

Reply:
Fixed.